iocuddle/src/lib.rs

```diff
@@ -0,0 +1,372 @@
+#[repr(transparent)]
+pub struct Group(u8);
+
+impl Group {
+    /// Create a new group for related `ioctl`s from its allocated number
+    pub const fn new(value: u8) -> Self {
+        Self(value)
+    }
+
+    const unsafe fn make<'a, D, T>(self, nr: u8, dir: c_ulong) -> Ioctl<D, &'a T> {
+        const SIZE_BITS: c_ulong = 14;
+
+        let mut req = dir;
+
+        req <<= SIZE_BITS;
+        req |= core::mem::size_of::<T>() as c_ulong & ((1 << SIZE_BITS) - 1);
+
+        req <<= 8;
+        req |= self.0 as c_ulong;
+
+        req <<= 8;
+        req |= nr as c_ulong;
+
+        Ioctl::classic(req)
+    }
+
+    /// Define a new `Read` `ioctl` with an associated `type`
+    ///
+    /// The `nr` argument is the allocated integer which uniquely
+    /// identifies this `ioctl` within the `Group`.
+    ///
+    /// # Safety
+    ///
+    /// For safety details, see [Ioctl::classic].
+    pub const unsafe fn read<'a, T>(self, nr: u8) -> Ioctl<Read, &'a T> {
+        self.make(nr, 0b10)
+    }
+
+    /// Define a new `Write` `ioctl` with an associated `type`
+    ///
+    /// The `nr` argument is the allocated integer which uniquely
+    /// identifies this `ioctl` within the `Group`.
+    ///
+    /// # Safety
+    ///
+    /// For safety details, see [Ioctl::classic].
+    pub const unsafe fn write<'a, T>(self, nr: u8) -> Ioctl<Write, &'a T> {
+        self.make(nr, 0b01)
+    }
+
+    /// Define a new `WriteRead` `ioctl` with an associated `type`
+    ///
+    /// The `nr` argument is the allocated integer which uniquely
+    /// identifies this `ioctl` within the `Group`.
+    ///
+    /// # Safety
+    ///
+    /// For safety details, see [Ioctl::classic].
+    pub const unsafe fn write_read<'a, T>(self, nr: u8) -> Ioctl<WriteRead, &'a T> {
+        self.make(nr, 0b11)
+    }
+}
+
+/// A defined `ioctl` along with its associated `direction` and `type`
+#[derive(Copy, Clone, PartialEq, Eq)]
+#[repr(transparent)]
+pub struct Ioctl<D, T>(c_ulong, PhantomData<(D, T)>);
+
+impl<D, T> Ioctl<D, T> {
+    /// Define a new `ioctl` with an associated `direction` and `type`
+    ///
+    /// The `request` argument is the allocated integer which uniquely
+    /// identifies this `ioctl`.
+    ///
+    /// # Safety
+    ///
+    /// This function is unsafe because defining an `ioctl` with an incorrect
+    /// `request`, `direction` or argument `type` can later result in memory
+    /// corruption. You are responsible to ensure these values are correct.
+    ///
+    /// Further, you are responsible to ensure that the argument `type` itself
+    /// provides appropriate safe wrappers around its raw contents. For some
+    /// `type`s none are required. For others, particularly `type`s that pass
+    /// pointers to the kernel as `u64`, you need to ensure that things like
+    /// lifetimes are correct.
+    pub const unsafe fn classic(request: c_ulong) -> Self {
+        Self(request, PhantomData)
+    }
+}
+
+impl Ioctl<Read, c_void> {
```

Comment:
Could we avoid having to use `c_void` by making the second argument an `Option` and pass `None` here?

Reply:
No. `None` is not a type. We need a type in order to mark this as having no parameter. The type that probably makes the most sense is the never type. But it isn't stable. `c_void` can't be constructed and has a long history being used as "no argument" in C (for example: `int foo(void);`).

https://doc.rust-lang.org/std/primitive.never.html

Reply:
That makes sense, thanks!